src/fe-common/core/fe-windows.c

```diff
@@ -84,30 +88,60 @@ WINDOW_REC *window_create(WI_ITEM_REC *item, int automatic)
 	return rec;
 }
 
+static void window_set_refnum0(WINDOW_REC *window, int refnum)
+{
+	int old_refnum;
+
+	g_return_if_fail(window != NULL);
+	g_return_if_fail(refnum >= 1);
+	if (window->refnum == refnum) return;
+
+	old_refnum = window->refnum;
+	window->refnum = refnum;
+	signal_emit("window refnum changed", 2, window, GINT_TO_POINTER(old_refnum));
+}
+
+static int window_refnum_lookup(WINDOW_REC *window, void *refnum_p)
+{
+	int refnum = GPOINTER_TO_INT(refnum_p);
+	return window->refnum == refnum ? 0 : window->refnum < refnum ? -1 : 1;
+}
+
 /* removed_refnum was removed from the windows list, pack the windows so
    there won't be any holes. If there is any holes after removed_refnum,
    leave the windows behind it alone. */
 static void windows_pack(int removed_refnum)
 {
 	WINDOW_REC *window;
 	int refnum;
+	GSequenceIter *iter, *end;
+
+	refnum = removed_refnum+1;
+	end = g_sequence_get_end_iter(windows_seq);
+	iter = g_sequence_lookup(windows_seq, GINT_TO_POINTER(refnum), (GCompareDataFunc)window_refnum_lookup, NULL);
+	if (iter == NULL) iter = end;
 
-	for (refnum = removed_refnum+1;; refnum++) {
-		window = window_find_refnum(refnum);
-		if (window == NULL || window->sticky_refnum)
+	for (; iter != end; refnum++, iter = g_sequence_iter_next(iter)) {
+		window = g_sequence_get(iter);
+
+		if (window == NULL || window->sticky_refnum || window->refnum != refnum)
```

Comment:
And can `g_sequence_get` return `NULL`, if it does I guess something went wrong and an `assert` should be used instead.